modules/util/source-formatter/src/main/java/com/liferay/source/formatter/check/UpgradeVelocityIfStatementsMigrationCheck.java

```diff
@@ -0,0 +1,153 @@
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.source.formatter.check;
+
+import com.liferay.petra.string.CharPool;
+import com.liferay.petra.string.StringPool;
+import com.liferay.portal.kernel.util.StringUtil;
+
+import java.util.Stack;
+
+/**
+ * @author Nícolas Moura
+ */
+public class UpgradeVelocityIfStatementsMigrationCheck
+	extends BaseUpgradeVelocityMigrationCheck {
+
+	@Override
+	protected String migrateContent(String content) {
+		String[] lines = content.split(StringPool.NEW_LINE);
+
+		for (int i = 0; i < lines.length; i++) {
+			String line = lines[i];
+
+			String newLine = line;
+
+			boolean parenthesis = false;
+
+			if (line.contains(_VELOCITY_IF_START) &&
+				_isVelocityStatement(line, _VELOCITY_IF_START)) {
+
+				parenthesis = true;
+
+				newLine = StringUtil.replace(
+					newLine, _VELOCITY_IF_START, "<#if");
+
+				_replaceIfStatementEnd(lines, i);
+			}
+			else if (line.contains(_VELOCITY_ELSEIF_START) &&
+					 _isVelocityStatement(line, _VELOCITY_ELSEIF_START)) {
+
+				parenthesis = true;
+
+				newLine = StringUtil.replace(
+					newLine, _VELOCITY_ELSEIF_START, "<#elseif");
+			}
+			else if (line.contains(_VELOCITY_ELSE_START) &&
+					 !line.contains(_VELOCITY_ELSEIF_START) &&
+					 _isVelocityStatement(line, _VELOCITY_ELSE_START)) {
+
+				newLine = StringUtil.replace(
+					newLine, _VELOCITY_ELSE_START, "<#else>");
+			}
+
+			if (parenthesis) {
+				if (_hasBreakLine(newLine)) {
+					String nextLine = lines[i + 1];
+
+					lines[i + 1] = StringUtil.replaceLast(
+						nextLine, CharPool.CLOSE_PARENTHESIS,
+						CharPool.GREATER_THAN);
+				}
+				else {
+					newLine = StringUtil.replaceLast(
+						newLine, CharPool.CLOSE_PARENTHESIS,
+						CharPool.GREATER_THAN);
+				}
+
+				newLine = StringUtil.removeFirst(
+					newLine, StringPool.OPEN_PARENTHESIS);
+			}
+
+			lines[i] = newLine;
+		}
+
+		return com.liferay.petra.string.StringUtil.merge(
+			lines, StringPool.NEW_LINE);
+	}
+
+	private static void _replaceIfStatementEnd(String[] lines, int lineIndex) {
+		Stack<String> stack = new Stack<>();
+
+		stack.push(_VELOCITY_IF_START);
+
+		int nextLineIndex = lineIndex;
+
+		while (!stack.empty()) {
+			nextLineIndex += 1;
+
+			String nextLine = lines[nextLineIndex];
+
+			if (nextLine.contains(_VELOCITY_IF_START) &&
+					_isVelocityStatement(nextLine, _VELOCITY_IF_START)) {
+				stack.push(_VELOCITY_IF_START);
+			}
+
+			if (nextLine.contains(_VELOCITY_FOREACH_START)) {
+				stack.push(_VELOCITY_FOREACH_START);
+			}
+
+			if (nextLine.contains(_VELOCITY_MACRO_START) &&
+					_isVelocityStatement(nextLine, _VELOCITY_MACRO_START)) {
+				stack.push(_VELOCITY_MACRO_START);
+			}
+
+			if (nextLine.contains(_VELOCITY_END)) {
+				stack.pop();
+			}
+
+			if (stack.empty()) {
+				lines[nextLineIndex] = StringUtil.replace(
+					nextLine, _VELOCITY_END, "</#if>");
+			}
+		}
+	}
```

Comment:
Hi @kevhlee, this is the method that I'm thinking in extract to a Util class on the next check. It is a piece of code that I will need to use in the If, #foreach, and #macro checks. These three will need this logic to replace only the correct #end because in velocity all statements end with #end but in FreeMarker they don't, so it needs to be replaced correctly.

It would only be necessary to add one parameter receiving the type of statement you want to replace to generalize this method to be used by all three.

What do you think about this? I thought of a Utility class because only these three checks are going to use it, if all three were going to use it it would make more sense to add it to the Base class, but that is not the case.